daft/io/lance/lance_scan.py

```
@@ -72,8 +72,9 @@ def to_scan_tasks(self, pushdowns: PyPushdowns) -> Iterator[ScanTask]:
                 else pushdowns.columns + filter_required_column_names
             )
 
-        # TODO: figure out how to translate Pushdowns into LanceDB filters
-        filters = None
+        from daft.expressions import Expression
+
+        filters = Expression._from_pyexpr(pushdowns.filters).to_arrow_expr() if pushdowns.filters is not None else None
```

what happens if the filter expression cannot be converted to an arrow expression? Will this raise an error or fail silently?

---

The `to_arrow_expr` function will raise an error if it cannot be converted.

---

so I think for this to _properly_ work, we'd need to modify the `can_absorb_filter()` to have the filter expression as an argument so we can check if it is a filter that we can actually absorb. 

```py
def can_absorb_filter(self, expr):
  try:
    Expression._from_pyexpr(expr).to_arrow_expr()
    return true
  except: # catch the proper exception here
    return false
```

Without a check like this, I'm afraid that the scan operation will fail at runtime if we pass in an expression that can't be converted to arrow.

---

@universalmind303 Great suggestion. However, my advice here is to create another PR. Because the current pushdown expression can only come from to_scan_tasks and cannot appear in the methods of ScannerOperator. I think I should do the check of this pushdown in another PR that binds the scanoperator to pushdown? This PR is specifically for solving the problem of lance expression conversion? WDYT?

---

Since this is best-effort, and we do not have pushdown support today, I suggest doing a best effort - try to convert the expression, if it fails, then don't push it down. @colin-ho and I discussed this a few months back now, and I believe you can have `can_absorb_filter` always `True`, and even if you don't apply the predicate, daft will anyways. Colin, please feel free to correct me if anything has changed. I believe @Jay-ju could confirm by explaining plans and will find a `filter` above the scan.

---

@rchowell https://github.com/Eventual-Inc/Daft/pull/4612. Actually, in this PR, I'm trying to make the explain really based on whether pushdown is supported for display. Currently, the plan that can't be pushed down is also shown in the pushdown parameter of the scanner. If needed, I'll separate the display of pushdown. Because in that PR, there is also the display of limit, but many UTs didn't pass.

---

Hey @Jay-ju just two things for you to confirm/test and then we can move ahead!

**(1) set `can_absorb_limit` to `False`**

I don't believe it's safe to set the `can_absorb_limit`, and I have some notes[ I took at the time](https://github.com/Eventual-Inc/Daft/pull/4243) which should explain why.

```python

    def can_absorb_limit(self) -> bool:
        """The `can_absorb_limit` toggles limit pushdowns.
        If this is true, then the scan operator MUST apply the limit.
        Daft will include always include a limit in the pushdown if possible,
        regardless of this value. This allows scans to apply a local limit
        to each of its tasks, but daft will still handle the global limit.
        In this example, I am unable to apply a global limit to the scan
        without materializing the datasource in `to_scan_tasks` which is
        not ideal .. so instead I will pass the limit to each ScanTask and
        apply locally, then daft will handle the global limit for correctness.
        """
        return False
```

**(2) Ensure you update the return schema to match EXACTLY the updated schema from any applied projects**

See in the example that I must update the schema if there are any projections!

```python
        # PROJECTION PUSHDOWN
        #
        #   This one is easy since pyarrow can accept a list of strings
        #   as well as indexes into the schema. We can't push down projections
        #   of arbitrary expressions because we don't have any pre-computed
        #   columns. On one hand, we could walk the terms, but we actually
        #   already have columns as strings via the PyPushdowns rust wrapper.
        #   So we grab that directly and don't need to use terms.
        #
        #   option (1) using existing column names
        #
        # projections: list[str] | None = pushdowns.columns
        #
        #   option (2) extract columns for arbitrary expressions
        #
        if pd.projections:
            schema = _translate_projections(self._schema, pd.projections)
            projections = schema.column_names()
        else:
            schema = self._schema
            projections = None
```

Then we use this schema in the scan tasks. 

```python
for task in [TestScanTask(frag, predicate, projections, limit) for frag in self._ds.get_fragments()]:
            yield ScanTask.python_factory_func_scan_task(
                module=_apply_task.__module__,
                func_name=_apply_task.__name__,
                func_args=(task,),
                schema=schema._schema,  # <-- USE UPDATED SCHEMA !!
                num_rows=None,
                size_bytes=None,
                pushdowns=None,
                stats=None,
            )
```

This _may_ already be the case today, but could you please add tests and confirm, otherwise projection pushdowns will trigger internal errors on schema mismatch. Thank you!

---

> I don't believe it's safe to set the `can_absorb_limit`, and I have some notes[ I took at the time](https://github.com/Eventual-Inc/Daft/pull/4243) which should explain why.

@rchowell  I don't understand why it's considered unsafe here. All I did was push down the local limit for the scanner. That is, if the scan operator has 10 tasks, each task will only output a limited number of data entries. Then, a global limit is used to output the overall limited data. Here, can_absorb_limit should be used to indicate whether the local limit can be connected to the data source. I don't see where the problem lies.

---

I believe `can_absorb_limit` indicates the **can absorb the global limit** which is why it is not safe. If you have tests to prove otherwise, I believe you! I haven't revisited in months, and I left those detailed notes specifically because I was impacted by this issue.

---

> I believe `can_absorb_limit` indicates the **can absorb the global limit** which is why it is not safe. If you have tests to prove otherwise, I believe you! I haven't revisited in months, and I left those detailed notes specifically because I was impacted by this issue.

I'm quite curious about this. Why does `can_absorb_limit` indicate "can absorb the global limit"? Does "scanner" here mean "global"? If this option is enabled, will there be no global limit at the planning level?

---

To be honest, I don't know exactly why, which is why I suggest testing. I believe that if you enable this option, then NO global limit will be applied at the planning level.

However, if you set this option to `False`, then a local limit is still passed in AND the plan has a global limit. You should do it this way. These were my findings months back when I implemented a ScanOperator and tested all the pushdown logic.

---

@rchowell 
1. limit I checked the code and indeed, `can_absorb_limit` determines whether to remove the global limit. This is not in line with expectations, so I've reverted the logic of limit here.
2. The project can be obtained through the expression currently, so there is no problem with this.
3. The logic of filter needs to be modified. When filter push - down is not supported, the current optimizer still shows that it is pushed down to the datasource.